src/blockchain/mod.rs

```diff
@@ -164,6 +164,72 @@ pub trait ConfigurableBlockchain: Blockchain + Sized {
     fn from_config(config: &Self::Config) -> Result<Self, Error>;
 }
 
+/// Trait for blockchains that don't contain any state
+///
+/// Statless blockchains can be used to sync multiple wallets with different descriptors.
+///
+/// [`BlockchainFactory`] is automatically implemented for `Arc<T>` where `T` is a stateless
+/// blockchain.
+pub trait StatelessBlockchain: Blockchain {}
+
+/// Trait for a factory of blockchains that share the underlying connection or configuration
+#[cfg_attr(
+    not(feature = "async-interface"),
+    doc = r##"
+## Example
+
+This example shows how to sync multiple walles and return the sum of their balances
+
+```no_run
+# use bdk::Error;
+# use bdk::blockchain::*;
+# use bdk::database::*;
+# use bdk::wallet::*;
+# use bdk::*;
+fn sum_of_balances<B: BlockchainFactory>(blockchain_factory: B, wallets: &[Wallet<MemoryDatabase>]) -> Result<u64, Error> {
+    Ok(wallets
+        .iter()
+        .map(|w| -> Result<_, Error> {
+            let wallet_name = wallet_name_from_descriptor(
+                w.public_descriptor(KeychainKind::External)?.unwrap(),
+                w.public_descriptor(KeychainKind::Internal)?,
+                w.network(),
+                w.secp_ctx()
+            )?;
+            w.sync(&blockchain_factory.build(&wallet_name, None)?, SyncOptions::default())?;
+            w.get_balance()
+        })
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .sum())
+}
+```
+"##
+)]
+pub trait BlockchainFactory {
+    /// The type returned when building a blockchain from this factory
+    type Inner: Blockchain;
+
+    /// Build a new blockchain for the given descriptor wallet_name
+    ///
+    /// If `override_skip_blocks` is `None`, the returned blockchain will inherit the number of blocks
+    /// from the factory. Since it's not possible to override the value to `None`, set it to
+    /// `Some(0)` to rescan from the genesis.
+    fn build(
+        &self,
+        wallet_name: &str,
+        override_skip_blocks: Option<u32>,
+    ) -> Result<Self::Inner, Error>;
```

Comment:
Does it make sense to add auto syncing in the `build` function? So if I call `factory.build_for_wallet()` then it will take the wallet database and perform and auto sync/setup on first instantiation??

Reply:
I wouldn't do it here in the `build()`, this is just to get a blockchain which you could use for anything, not necessarily just to sync a wallet.

We could add another method to the trait to automatically infer the wallet name from a wallet reference, and maybe even one more to sync a wallet without even returning the new blockchain. Just create it internally, sync, and throw it away.

Reply:
Unless we have a specific use case for why someone would want to sync a wallet and not keep the blockchain here I'd say best not to add any new functionality.  But if a specific case comes up where it's needed it can always be added in a future PR.